modules/soc_i3/i3soc.py

```diff
@@ -0,0 +1,210 @@
+    try:
+        url = 'https://' + auth_server + '/gcdm/oauth/authenticate'
+        headers = {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 15_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.3 Mobile/15E148 Safari/604.1'}
+        data = {
+            'client_id': client_id,
+            'response_type': 'code',
+            'scope': 'openid profile email offline_access smacc vehicle_data perseus dlm svds cesim vsapi remote_services fupo authenticate_user',
+            'redirect_uri': 'com.bmw.connected://oauth',
+            'state': state,
+            'nonce': 'login_nonce',
+            'code_challenge': code_challenge,
+            'code_challenge_method': 'plain',
+            'authorization': authcode1}
+        cookies = {
+            'GCDMSSO': authcode1}
+      
+        response = postHTTP(url, data, headers, cookies, allow_redirects=False)
+        authcode = dict(urllib.parse.parse_qsl(response.split("?",1)[1]))["code"]
+    except:
+        raise
+        
+    return authcode
+    
+def authStage3(authcode2, code_challenge):
+    try:
+        url = 'https://' + auth_server + '/gcdm/oauth/token'
+        headers = {
+            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
+            'Authorization': create_auth_string(client_id, client_password)}
+        data = {
+            'code': authcode2,
+            'code_verifier': code_challenge,
+            'redirect_uri': 'com.bmw.connected://oauth',
+            'grant_type': 'authorization_code'}
+      
+        response = postHTTP(url, data, headers, allow_redirects=False)
+        token = json.loads(response)
+    except:
+        raise
+        
+    return token
+    
+def requestToken(username, password):
+    try:
+        code_challenge = get_random_string(86)
+        state = get_random_string(22)
+    
+        authcode1 = authStage1(username, password, code_challenge, state)
+        authcode2 = authStage2(authcode1, code_challenge, state)
+        token = authStage3(authcode2, code_challenge)
+    except:
+        raise
+        
+    return token
+
+# ---------------Interface Function-------------------------------------------
+def requestData(token, vin):
+    url = 'https://' + api_server + '/eadrax-vcs/v1/vehicles?apptimezone=0&appDateTime=' + str(int(time.time())) + '&tireGuardMode=ENABLED'
+    headers = {
+          'x-user-agent': 'android(v1.07_20200330);bmw;1.7.0(11152)',
+          'Authorization': (token["token_type"] + " " + token["access_token"])}
+    response = json.loads(getHTTP(url, headers))
+    
+    for data in response:    
+        if data["vin"] == vin:
+            return data
+    
+    return 0
+    
+# ---------------Main Function-------------------------------------------
+def main():
+    try:
+        argsFile=str(sys.argv[1])
+        f = open(argsFile, 'r')

```

Comment:
Ressourcen sollten immer mit einem Kontexthandler geöffnet werden. In diesem Fall:
```
with open(argsFile, 'r') as f:
    argsStr = f.read()
argsDict = json.loads(argsStr)
```